src/daemon/gnet/pool.go

```diff
@@ -239,43 +239,56 @@ func (pool *ConnectionPool) Run(q chan struct{}) {
 
 // Shutdown gracefully shutdown the connection pool
 func (pool *ConnectionPool) Shutdown() {
+	pool.strand(func() error {
+		pool.addresses = map[string]*Connection{}
+		pool.pool = map[int]*Connection{}
+		return nil
+	})
+
 	close(pool.quit)
 	pool.listener.Close()
 	pool.listener = nil
-	pool.pool = make(map[int]*Connection)
-	pool.addresses = make(map[string]*Connection)
 }
 
 // strand ensures all read and write action of pool's member variable are in one thread.
-func (pool *ConnectionPool) strand(f func()) {
+func (pool *ConnectionPool) strand(f func() error) (err error) {
+	defer func() {
+		// send on closed operation channel will panic.
+		if r := recover(); r != nil {
+			err = ErrConnectionPoolClosed
```

Comment:
No need to log the recover, the panic is caused by sending on closed operations channel, and the channel will be closed after the connection pool is shutdown. Every methods calls after the connection pool is closed will cause this panic, but what the caller need to know is the connection pool is closed, don't try again.